### DISCUSSION TITLE
[Question]: What does `INTERNAL_KIND` mean in python api?

### DISCUSSION THREAD

**Discussion by hagozaebii** [cvc5/cvc5#10436] [General]
Hello,

Currently, I am using `simplify` in cvc5 base API. I use the getOp function to figure out what operator is used in the current term. And when I use `simplify`, I often see `INTERNAL_KIND`. The document says `This kind serves as an abstraction for internal kinds that are not exposed via the API but may appear in terms returned by API functions, e.g., when querying the simplified form of a term.`. Even though I read the document, I don't know what it means. What role does this play? Also, when `INTERNAL_KIND` is received as the return value of getOp, how to find out the real operator of that term?

To summarize my questions.
1. What is `INTERNAL_KIND`'s role?
2. How can I know the real operator when I get `INTERNAL_KIND` as a result of `getOp` method?

**Comment by ajreynol** [comment]
Internally, cvc5 introduces many operators and variables for solving that are not available in user inputs.  The kinds of these terms is known internally, but intentionally not exposed in the API.  Any term of this form will report `INTERNAL_KIND` as its kind.

Generally, any term with `INTERNAL_KIND` should not be used as input to our API, since we do not expect to see these terms as inputs to user-facing methods.

We are working on exposing more of the details of any term we export in the API.  In particular, this PR exports identifiers for all internal variables we introduce that may appear in user output: https://github.com/cvc5/cvc5/pull/10393.  

Can you share the term you are simplifying?  We can see if the particular operators you are seeing can be exported as part of this effort.

**Comment by hagozaebii** [comment]
@ajreynol First of all, Thank for your kindness always.

I understood your answer to mean that there is currently no way to know the operator when `INTERNAL_KIND` appears. Is it right?
And, the term that I simplified is as follows:
```
(not (str.suffixof (ite (str.in_re (ite (=> true false) (str.at var_9 6) (str.update var_10 8 var_10)) (re.* (re.++ re.allchar re.allchar))) (str.from_code (- 3)) (str.replace_all (str.to_upper (str.from_code 5)) (str.replace_re_all (str.to_upper var_9) (re.range "C" "H") var_10) (str.++ (str.to_upper var_10) (str.from_code 2)))) “z”)”
```
Thank you again!

**Comment by ajreynol** [comment]
Yes, the intention is that terms with `INTERNAL_KIND` are intentionally left unspecified to the user; the fix is to ensure that no terms with this kind are ever given back to the user so that we have "official" support for all terms.  

FYI, the operator in question here is the (internal, total version of) integer modulus. 

I will discuss this with the cvc5 team and we will prioritize addressing these cases. As I mentioned, https://github.com/cvc5/cvc5/pull/10393 is one step in this direction.

**Comment by ajreynol** [comment]
FYI, I've opened https://github.com/cvc5/cvc5/pull/10460 which should address this instance.
